File: leetcode/20.py
class Solution:
    def isValid(self, s: str) -> bool:
        # similar to stacks
        # add 1 to a placeholder value, 
        # need to check if ) happens before (
        # make a new list, if ( is in our values, add itto the list, if there 
        # is a ), then pop the value. if there is no previous ) then it should
        # return an error 
        stack = []
        brackets = {")": "(", "}": "{", "]": "["}
        
        for char in s:
            if char in brackets.values():  
                stack.append(char)
            elif char in brackets: 
                if not stack or stack[-1] != brackets[char]:
                    return False
                stack.pop()
            else:
                return False
        
        return not stack  

File: leetcode/20.cpp
#include <iostream>
#include <stack>
#include <string>

class Solution {
public:
    bool isValid(std::string s) {
        std::stack<char> stack;

        for (int i = 0; i < s.size(); i++) {
            if (s[i] == '{' || s[i] == '[' || s[i] == '(') {
                stack.push(s[i]);
            } else {
                if (stack.empty()) {
                    return false;
                }
                if ((s[i] == '}' && stack.top() != '{') ||
                    (s[i] == ']' && stack.top() != '[') ||
                    (s[i] == ')' && stack.top() != '(')) {
                    return false;
                }
                stack.pop();
            }
        }

        return stack.empty();
    }
};